libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpMPEG4Reader.java

```diff
@@ -0,0 +1,141 @@
+import androidx.media3.extractor.ExtractorOutput;
+import androidx.media3.extractor.TrackOutput;
+import com.google.common.primitives.Bytes;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+
+/**
+ * Parses an MPEG4 byte stream carried on RTP packets, and extracts MPEG4 Access Units. Refer to
+ * RFC6416 for more details.
+ */
+/* package */ final class RtpMPEG4Reader implements RtpPayloadReader {
+  private static final String TAG = "RtpMPEG4Reader";
+
+  private static final long MEDIA_CLOCK_FREQUENCY = 90_000;
+
+  /**
+   * VOP unit type.
+   */
+  private static final int I_VOP = 0;
+
+  private final RtpPayloadFormat payloadFormat;
+  private @MonotonicNonNull TrackOutput trackOutput;
+  @C.BufferFlags private int bufferFlags;
+  private long firstReceivedTimestamp;
+  private int previousSequenceNumber;
+  private long startTimeOffsetUs;
+  private int sampleLength;
+
+  /** Creates an instance. */
+  public RtpMPEG4Reader(RtpPayloadFormat payloadFormat) {
+    this.payloadFormat = payloadFormat;
+    firstReceivedTimestamp = C.TIME_UNSET;
+    previousSequenceNumber = C.INDEX_UNSET;
+    sampleLength = 0;
+  }
+
+  @Override
+  public void createTracks(ExtractorOutput extractorOutput, int trackId) {
+    trackOutput = extractorOutput.track(trackId, C.TRACK_TYPE_VIDEO);
+    castNonNull(trackOutput).format(payloadFormat.format);
+  }
+
+  @Override
+  public void onReceivingFirstPacket(long timestamp, int sequenceNumber) {
+    Log.i(TAG, "RtpMPEG4Reader onReceivingFirstPacket");
+  }
+
+  @Override
+  public void consume(ParsableByteArray data, long timestamp, int sequenceNumber, boolean rtpMarker)
+      throws ParserException {
+    if (previousSequenceNumber != C.INDEX_UNSET && sequenceNumber != (previousSequenceNumber + 1)) {
+      Log.e(TAG, "Packet loss");
+    }
+    checkStateNotNull(trackOutput);
+
+    int limit = data.bytesLeft();
+    trackOutput.sampleData(data, limit);
+    sampleLength += limit;
+    parseVopType(data);
+
+    // Marker (M) bit: The marker bit is set to 1 to indicate the last RTP
+    // packet(or only RTP packet) of a VOP. When multiple VOPs are carried
+    // in the same RTP packet, the marker bit is set to 1.
+    if (rtpMarker) {
+        if (firstReceivedTimestamp == C.TIME_UNSET) {
+          firstReceivedTimestamp = timestamp;
+        }
+
+        long timeUs = toSampleUs(startTimeOffsetUs, timestamp, firstReceivedTimestamp);
+        trackOutput.sampleMetadata(timeUs, bufferFlags, sampleLength, 0, null);
+        sampleLength = 0;
+      }
+
+    previousSequenceNumber = sequenceNumber;
+  }
+
+  @Override
+  public void seek(long nextRtpTimestamp, long timeUs) {
+    firstReceivedTimestamp = nextRtpTimestamp;
+    startTimeOffsetUs = timeUs;
+    sampleLength = 0;
+  }
+
+  // Internal methods.
+
+  /**
+   * Parses VOP Coding type
+   *
+   * Sets {@link #bufferFlags} according to the VOP Coding type.
+   */
+  private void parseVopType(ParsableByteArray data) {
```

Comment:
Since we are parsing the buffer flag here, I would suggest making this method
 - return the buffer flag, like you did with `getBufferFlagsFromVopType()`
 - a private static method